### PR TITLE
Refactor TypeRegistry

### PR DESCRIPTION
**Pull request by arhimondr**

### CODE REVIEW

**Comment by martint** [comment]
There are test failures. Can you take a look?

**Comment by haozhun** [comment]
@arhimondr Is there a particular reason we are moving functions to operate on `Type`s instead of `TypeSignature`s? That seems to be opposite from the general direction we want to move.

**Comment by arhimondr** [comment]
@martint Tests are failing on master as well. I think this patch (https://github.com/prestodb/presto/commit/e77fa0696c39db9b4fd0959ef8eb3aaa9708c8f2) causing a failure. We are trying to insert more columns than some table has. I can create a patch fixing that.

@haozhun `CanCoerce` and `GetCommonSupperType` methods that operate on signatures are really confusing. `TypeSignature` can be parametrized, like `Array<T>` or `Decmal(x,y)`. So, what should be returned by `canCoerce` or `getCommonSupperType` for the following parameters pair? 

- `Array<T>`, `UNKNOWN`
- `Array<T>`,`Array<BIGINT>`
- `Array<BIGINT>`,`Array<T>`
- `DECIMAL(38,10)`,`DECIMAL(x,y)`

I think any answer wont be intuitive. 

Those method that accept `TypeSignature` were only needed by `SignatureBinder` that is used in `FunctionRegistry` for function resolution. So, i though it is better to handle signature only there.

**Comment by yuananf** [comment]
@martint @arhimondr it seems #5176 fix the Travis

**Comment by arhimondr** [comment]
@yuananf Yes, it fixes failing integration tests.

**Comment by haozhun** [comment]
@arhimondr The way I think about it: `Type` is a heavy-weight extension around `TypeSignature` and provide additional features that the latter doesn't provide.
* 2 helper methods: `getDisplayName`, `getTypeParameters`
* 2 additional properties: `isComparable`, `isOrderable`
* methods for operating on Blocks: `getJavaType`, `createBlockBuilder`, `createBlockBuilder`, `getObjectValue`, `getBoolean`, `getLong`, `getDouble`, `getSlice`, `getObject`, `writeBoolean`, `writeLong`, `writeDouble`, `writeSlice`, `writeSlice`, `writeObject`, `appendTo`
* semi-deprecated methods: `equalTo`, `hash`, `compareTo`

You are right that `TypeSignature` can work on `ParameterKind` of `VARIABLE` while `Type` can't. But it's okay to write functions that takes `TypeSignature` yet doesn't handle `ParameterKind` of `VARIABLE`

**Comment by arhimondr** [comment]
@haozhun From my point of view `Type` is something that is physically exists. User can enter value of some `Type`, some connector can provide tables with the column of some `Type`, etc. `TypeSignature` is not a `Type`. It is just an abstract signature of something. I agree with you that `Type` provides a lot of extra methods that are not needed during the `canCoerce` calculations, but i think that operating on `Type` in that methods makes clear that such methods can operate only on real types. 

Also the code of those methods become cleaner. Instead of handling every parameter in type signature, and putting a lot of assertions  we can just use public methods of some concrete type. Let me show you an example with `DECIMAL`

On `TypeSignature`s

```
if (firstType.getBase().equals(DECIMAL) && secondType.getBase().equals(DECIMAL)) {
            checkState(firstType.getParameters().size() == 2);
            checkState(secondType.getParameters().size() == 2);
            TypeSignatureParameter firstPrecisionParameter = firstType.getParameters().get(0);
            TypeSignatureParameter firstScaleParameter = firstType.getParameters().get(1);
            TypeSignatureParameter secondPrecisionParameter = secondType.getParameters().get(0);
            TypeSignatureParameter secondScaleParameter = secondType.getParameters().get(1);
            checkState(firstPrecisionParameter.isLongLiteral());
            checkState(firstScaleParameter.isLongLiteral());
            checkState(secondPrecisionParameter.isLongLiteral());
            checkState(secondScaleParameter.isLongLiteral());
            long firstPrecision = firstPrecisionParameter.getLongLiteral();
            long secondPrecision = secondPrecisionParameter.getLongLiteral();
            long firstScale = firstScaleParameter.getLongLiteral();
            long secondScale = secondScaleParameter.getLongLiteral();
            long targetScale = Math.max(firstScale, secondScale);
            long targetPrecision = Math.max(firstPrecision - firstScale, secondPrecision - secondScale) + targetScale;
            targetPrecision = Math.min(38, targetPrecision); //we allow potential loss of precision here. Overflow checking is done in operators.
            return Optional.of(new TypeSignature(
                    DECIMAL,
                    ImmutableList.of(TypeSignatureParameter.of(targetPrecision), TypeSignatureParameter.of(targetScale))));
        }
```


On `Type`s

```
 int targetScale = Math.max(firstType.getScale(), secondType.getScale());
        int targetPrecision = Math.max(firstType.getPrecision() - firstType.getScale(), secondType.getPrecision() - secondType.getScale()) + targetScale;
        //we allow potential loss of precision here. Overflow checking is done in operators.
        targetPrecision = Math.min(38, targetPrecision);
        return createDecimalType(targetPrecision, targetScale);
```

**Comment by haozhun** [comment]
Your comparison wasn't fair. I rewrote them below. (You don't need to do the checkState for `isLongLiteral` because `getLongLiteral` will do the exact same `checkState`. You need to check the types regardless of whether you are working with `Type` or `TypeSignature`

But I get your point and I agree that working with Type makes the code cleaner. But they are not as different as is in your comment.

We've both made our point. I'll defer this discussion to @martint.

```
if (firstType.getBase().equals(DECIMAL) && secondType.getBase().equals(DECIMAL)) {
    checkState(firstType.getParameters().size() == 2);
    checkState(secondType.getParameters().size() == 2);
    long firstPrecision = firstType.getParameters().get(0).getLongLiteral();
    long secondPrecision = secondType.getParameters().get(0).getLongLiteral();
    long firstScale = firstType.getParameters().get(1).getLongLiteral();
    long secondScale = secondType.getParameters().get(1).getLongLiteral();
    long targetScale = Math.max(firstScale, secondScale);
    long targetPrecision = Math.max(firstPrecision - firstScale, secondPrecision - secondScale) + targetScale;
    // we allow potential loss of precision here. Overflow checking is done in operators.
    targetPrecision = Math.min(38, targetPrecision);
    return new TypeSignature(
            DECIMAL,
            ImmutableList.of(TypeSignatureParameter.of(targetPrecision), TypeSignatureParameter.of(targetScale)));
}
```

```
if (firstType instanceof DecimalType && secondType instanceof DecimalType) {
    DecimalType firstDecimalType = (DecimalType) firstType;
    DecimalType secondDecimalType = (DecimalType) secondType;
    int targetScale = Math.max(firstDecimalType.getScale(), secondDecimalType.getScale());
    int targetPrecision = Math.max(firstDecimalType.getPrecision() - secondDecimalType.getScale(), secondDecimalType.getPrecision() - secondDecimalType.getScale()) + targetScale;
    // we allow potential loss of precision here. Overflow checking is done in operators.
    targetPrecision = Math.min(38, targetPrecision);
    return createDecimalType(targetPrecision, targetScale);
}
```

**Comment by arhimondr** [comment]
@haozhun I agree that my example isn't completely fair from the LOC point of view. But LOC is not what i wanted to emphasize. My point is that when you are operating on `TypeSignature` you need to know which parameter from parameters list means what. You need to check that this parameter is specified. You need to check that this parameter is of the right type. In other words you need to duplicate `Type` creation logic which is already implemented in `ParametricType` interface implementation.

**Comment by martint** [comment]
In the current code, `Type` is a mix between a fully bound type representation and the type implementation details (e.g., how to compute equality/hashcode, read/write values from blocks, etc). 

My plan is to, eventually, decouple these so that the analyzer/type inference engine/planner can operate on types symbolically, and only fetch the implementations when it needs to produce the physical plan and during code generation. A related goal is to be able to push all the type implementations into connectors (or have the ability to do it), much like functions/operators work.

File: presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java [hunk 1]
```diff
@@ -144,29 +140,46 @@ private boolean matchArguments(
     }
 
     private boolean bind(
-            TypeSignature expected,
-            Type actual,
+            TypeSignature expectedSignature,
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java]: Why change the names of these arguments?

**Comment by martint** [comment]
Merged, thanks!

**Pull request closed without merging**